### PR TITLE
restrict txpool gp threshold by rgp & add eth_minGasPrice

### PR DESCRIPTION
**Pull request by LeoGuo621**

### CODE REVIEW

File: zk/txpool/pool.go [hunk 1]
```diff
@@ -737,8 +737,10 @@ func (p *TxPool) validateTx(txn *types.TxSlot, isLocal bool, stateCache kvcache.
 		return UnsupportedTx
 	}
 
-	// Drop non-local transactions under our own minimal accepted gas price or tip
-	if !isLocal && uint256.NewInt(p.cfg.MinFeeCap).Cmp(&txn.FeeCap) == 1 {
+	// Drop transactions under our raw gas price suggested by default\fixed\follower gp mode
+	rgp := p.gpCache.GetLatestRawGP()
+
+	if !p.isFreeGas(txn.SenderID) && uint256.NewInt(rgp.Uint64()).Cmp(&txn.FeeCap) == 1 {
```
Comment (sieniven) [inline zk/txpool/pool.go:743]: Add X Layer comment

Comment (sieniven) [reply zk/txpool/pool.go:743]: We should shift this GP validation check into a function call, and define the function inside pool_xlayer.

File: cmd/rpcdaemon/commands/eth_api.go [hunk 2]
```diff
@@ -389,9 +390,11 @@ func NewEthAPI(base *BaseAPI, db kv.RoDB, eth rpchelper.ApiBackend, txPool txpoo
...
 	// Only Sequencer requires to calculate dynamic gas price periodically
 	// eth_gasPrice requests for the RPC nodes are all redirected to the Sequencer node (via zkevm.l2-sequencer-rpc-url)
-	if sequencer.IsSequencer() {
-		apii.runL2GasPricerForXLayer()
-	}
+	GasPricerOnce.Do(func() {
+		if sequencer.IsSequencer() {
+			apii.runL2GasPricerForXLayer()
+		}
+	})
```
Comment (sieniven) [inline cmd/rpcdaemon/commands/eth_api.go:397]: Maybe we can shift the runniing the runL2GasPricerForXLayer into eth backend Init function?

File: cmd/rpcdaemon/commands/eth_api_xlayer.go [hunk 3]
```diff
@@ -11,5 +17,57 @@ func (apii *APIImpl) GetGPCache() *GasPriceCache {
...
+	c.values[timestamp] = new(big.Int).Set(rgp)
+
+	// Remove values older than 5 minutes
+	cleanupTime := time.Now().Add(-cacheRGPDuration)
+	for t := range c.values {
+		if t.Before(cleanupTime) {
+			delete(c.values, t)
+		}
+	}
+}
```
Comment (sieniven) [inline cmd/rpcdaemon/commands/eth_api_xlayer.go:53]: Does not really make sense to define our TTL cache with this. We can use a way more optimal go package like bigcache, or even implement a fixed size LRU cache, and pass the capacity on runtime with the 5minutes / update period

File: cmd/rpcdaemon/commands/eth_system_xlayer.go [hunk 4]
```diff
@@ -71,8 +71,8 @@ func (api *APIImpl) getLatestBlockTxNum(ctx context.Context) (int, error) {
 	return len(b.Transactions()), nil
 }
 
-func (api *APIImpl) getGPFromTrustedNode() (*big.Int, error) {
-	res, err := client.JSONRPCCall(api.l2RpcUrl, "eth_gasPrice")
+func (api *APIImpl) getGPFromTrustedNode(method string) (*big.Int, error) {
+	res, err := client.JSONRPCCall(api.l2RpcUrl, method)
```
Comment (sieniven) [inline cmd/rpcdaemon/commands/eth_system_xlayer.go:75]: Nitpicking here but this method seems more like a client wrapper call. Dont think we should change this. The cleanest way is to shift "eth_gasPrice" out into a constant but not necessary

File: zk/txpool/pool_xlayer.go [hunk 5]
```diff
@@ -61,3 +63,7 @@ func (p *TxPool) isFreeClaimAddr(senderID uint64) bool {
 func (p *TxPool) SetGpCacheForXLayer(gpCache GPCache) {
 	p.gpCache = gpCache
 }
+
+func (p *TxPool) isFreeGas(senderID uint64) bool {
+	return p.isFreeClaimAddr(senderID)
+}
```
Comment (sieniven) [inline zk/txpool/pool_xlayer.go:69]: Dont quite catch the point of this function

**Comment by zjg555543** [comment]
Please add e2e test for the 2 funcitons.

File: cmd/rpcdaemon/commands/eth_api_xlayer.go [hunk 6]
```diff
@@ -11,5 +16,45 @@ func (apii *APIImpl) GetGPCache() *GasPriceCache {
...
 	// set default gas price
 	apii.gasCache.SetLatest(common.Hash{}, apii.L2GasPricer.GetConfig().Default)
+	apii.gasCache.SetLatestRawGP(apii.L2GasPricer.GetConfig().Default)
 	go apii.runL2GasPriceSuggester()
 }
+
+const cacheSize = 30 // Circular buffer size
+
+type RawGPCache struct {
+	values []*big.Int // Circular buffer
```
Comment (sieniven) [inline cmd/rpcdaemon/commands/eth_api_xlayer.go:26]: Better to switch to fixed size array
```suggestion
	value [cacheSize]*big.Int
```

File: cmd/rpcdaemon/commands/eth_api_xlayer.go [hunk 7]
```diff
@@ -11,5 +16,45 @@ func (apii *APIImpl) GetGPCache() *GasPriceCache {
...
+	// Add the new RGP to the circular buffer
+	c.values[c.head] = new(big.Int).Set(rgp)
+	c.head = (c.head + 1) % cacheSize
+}
+
+// GetMin returns the minimum RGP in the circular buffer
+func (c *RawGPCache) GetMin() (*big.Int, error) {
+	if c.values[0] == nil { // The buffer is empty
+		return nil, fmt.Errorf("no values in cache")
+	}
```
Comment (sieniven) [inline cmd/rpcdaemon/commands/eth_api_xlayer.go:49]: Code will work but better to not do the empty check like this

File: cmd/rpcdaemon/commands/eth_api_xlayer.go [hunk 8]
```diff
@@ -11,5 +16,45 @@ func (apii *APIImpl) GetGPCache() *GasPriceCache {
...
+
+	minRGP := big.NewInt(0).SetInt64(math.MaxInt64) // Initialize to maximum big.Int
+	for _, value := range c.values {
+		if value != nil && value.Cmp(minRGP) < 0 {
+			minRGP = value
+		}
+	}
```
Comment (sieniven) [inline cmd/rpcdaemon/commands/eth_api_xlayer.go:56]:
```suggestion
	emptyFlag := true
	minRGP := big.NewInt(0).SetInt64(math.MaxInt64) // Initialize to maximum big.Int
	for _, value := range c.values {
		if value == nil {
			continue
		}
		emptyFlag = false
		if value.Cmp(minRGP) < 0 {
			minRGP = value
		}
	}
	
	if empty {
		return nil, fmt.Errorf("no values in cache")
	}
```

File: zk/txpool/pool.go (hunk 1 above)
Comment (sieniven) [reply zk/txpool/pool.go:743]: Shift p.isFreeGas check and min rawGP check into pool_xlayer.go, and maintain the old code (for better maintaining with upstream repo)

File: zk/txpool/pool.go [hunk 9]
```diff
@@ -737,8 +737,10 @@ func (p *TxPool) validateTx(txn *types.TxSlot, isLocal bool, stateCache kvcache.
 		return UnsupportedTx
 	}
 
-	// Drop non-local transactions under our own minimal accepted gas price or tip
-	if !isLocal && uint256.NewInt(p.cfg.MinFeeCap).Cmp(&txn.FeeCap) == 1 {
+	// Drop transactions under our raw gas price suggested by default\fixed\follower gp mode
+	// X Layer
+	rgp := p.gpCache.GetLatestRawGP()
+	if !p.isFreeGas(txn.SenderID) && uint256.NewInt(rgp.Uint64()).Cmp(&txn.FeeCap) == 1 {
```
Comment (sieniven) [inline zk/txpool/pool.go:743]:
```suggestion
	// For X Layer
	err := p.validateTxForXLayer()
	if err != nil {
		return err
	}
	
	// For X Layer, we disable the default pool underprice validation check
	// Drop non-local transactions under our own minimal accepted gas price or tip
	// if !isLocal && uint256.NewInt(p.cfg.MinFeeCap).Cmp(&txn.FeeCap) == 1 {
```

File: cmd/rpcdaemon/commands/eth_api_xlayer.go [hunk 10]
```diff
@@ -11,5 +16,45 @@ func (apii *APIImpl) GetGPCache() *GasPriceCache {
 func (apii *APIImpl) runL2GasPricerForXLayer() {
 	// set default gas price
 	apii.gasCache.SetLatest(common.Hash{}, apii.L2GasPricer.GetConfig().Default)
+	apii.gasCache.SetLatestRawGP(apii.L2GasPricer.GetConfig().Default)
 	go apii.runL2GasPriceSuggester()
 }
+
+const cacheSize = 30 // Circular buffer size
```
Comment (sieniven) [inline cmd/rpcdaemon/commands/eth_api_xlayer.go:23]: Write some comments on how we determine cacheSize
```suggestion
// 1200sec (TTL) / 10sec (UpdatePeriod)
const cacheSize = 30
```

**Review by sieniven — APPROVED** [review]

**Review by zjg555543 — APPROVED** [review]

**Pull request merged**